olive/passes/olive_pass.py

```diff
@@ -338,7 +339,15 @@ def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassCon
     def _resolve_defaults(cls, config: Dict[str, Any], default_config: Dict[str, PassConfigParam]) -> Dict[str, Any]:
         """Resolve default values."""
         for key, value in config.items():
-            if value == PassParamDefault.DEFAULT_VALUE:
+            if isinstance(default_config[key].searchable_values, CategoricalValue):
```

Comment:
PR #1401 is the example of using this search parameter.

In simple words, the requirement is that "suggestions for a `Categorical` parameter are user input driven".

In more words, consider execution providers as search parameter. User can cherry-pick, say two, of the many supported execution providers. Search needs to be run on only the chosen ones i.e. two search points.

Another example, consider a search parameter like QLoRA rank that has to search on a few different rank values. At the moment, this is hardcoded to [16, 32, 64] in the pass implementation. There is no way for the user to run this pass for [128, 256, 512]. `CategoricalValue` supports this use case. 

Longer explanation with example -
`OrtSessionParamsTuning` takes a list of execution providers as input. For the search to work, each of those entry in that user provided list is a search point. `CategoricalValue` is sort of a placeholder which is replaced by 'Categorical` in olive_pass.py::342. There are other such config parameters.

Reply:
so, what this does is tell the pass config that whatever the user provides, make it a categorical? This feels hacky and it also ignores the `disable_search` flag since it always becomes a search parameter. Is this expected? A no-search workflow will be invalid if perf-tuning pass is present? 

I think it would be better if the categoricalvalue change is part of the perf tuning PR so that the logic can be reviewed with it's usage. The way it is now, I cannot follow the logic and review/approve the changes related to it.

Reply:
if my understanding is correct, I think the `categoricalvalue` behavior might need to a separate flag in the passparamconfig. searchable_value is only used to hold the default search param when search is enabled or user asks for it using `SEARCHABLE_VALUES`. What you want looks more like a config validation to convert values to search parameter.

Reply:
cmiiw, what we want is an easy (maybe automatic?) way for a user to provide search values without having to configure it using something like `{"olive_parameter_type": "SearchParameter", "type": "Categorical", "support": [1, 2, 3]}`. The param can be configured directly using `[1,2,3]`. 

I think one way to achieve this would be to add a validator to the pass config class if search is enabled. this validator could convert all (or only some flagged paramters if we add such a flag) non-search parameters into categoricals. this is fine since `Categorical([1])` is the same a providing `1`. 
However, a potential issue for this is how do we recognize if it's a fixed value or searchable list if the type of the param is also a list/sequence? Maybe a flag would help with that so that the flag is only allowed on params whose type is not a sequence.

Reply:
>  The param can be configured directly using [1,2,3]

Yes, that's correct and that's what current implementation supports.

> ... non-search parameters into categoricals. this is fine since Categorical([1]) is the same as providing 1

I don't think that works. We need to convert searchable parameters to `Categorical`. How does changing non-searchable parameters help (yes `Categoriclas([1])` is the same as `1` but that's a moot point). The non-searchable `Categorical`s will get converted to constants again down the pipeline giving us the same config as before the change. It has absolutely no affect on the searchable parameters.

The problem is under certain conditions we want to treat user provided input as suggestions for `Categorical`, overriding the default, if present. Yes, we can do that swap/replacement as part of validation or somewhere else.

Reply:
by "non-search parameters`, I meant values that are provided like `1`, `[1,2]` instead of using `SearchParameter` object or json. There is nothing stopping a parameter from search and making it "non-searchable". Any parameter that has a search parameter as its value is "searchable". having default_value or default_searchable_value is not meant to be a marker for whether a param is searchable or not. they are just defaults that get used when search is enabled/disabled. 

what i was suggesting is that, if we want to convert user inputs to search params so that they don't need to fully specify a search parameter, we can just convert the values (non searchparameter types) into categorical. this works even if the user provided a single fixed value because of the moot point.
This way, the peft tuning pass is not a special case. we still have to tackle the list type issue regardless.

Reply:
Wouldn't such a solution be pass specific unless we add a new flag to the `PassConfigParam` and handle it generically in the validation logic.

Reply:
Yes, I think this is better done in the validation logic. We probably need a flag to enable such behavior per param using a flag because of the following issue otherwise:
> However, a potential issue for this is how do we recognize if it's a fixed value or searchable list if the type of the param is also a list/sequence? Maybe a flag would help with that so that the flag is only allowed on params whose type is not a sequence.

Not directly related, even this method is better handled in the validation logic https://github.com/microsoft/Olive/blob/c473a3b73458d55ff5d0da5fe0464338f258e824/olive/passes/olive_pass.py#L338

Reply:
Each parameter has three values.
A) User provided input value.
B) Default value selected by the pass implementor.
C) Searchable values selected by the pass implementer.

Today,
- If search is not enabled then A or B is used.
- If search is enabled and A is not provided, then C is used to generate search points. If A is provided than this parameter is not searched.

The purpose of this work is to provide the pass implementer new (if we don't already support it today) ability to support the following scenario where input type is a list.
- If search is not enabled then A or B is used.
- If search is enabled and A is not provided, then C is used to generate search points. If A is provided than A should be used to generate search points.

The pass implementor should be able to select how they want to treat the parameter. Make A searchable or not.